File: lc_plusplus/Copy List with Random Pointer.cc
// A linked list is given such that each node contains an additional random pointer which could point to any node in the list or null.
//
// Return a deep copy of the list.


/**
 * Definition for singly-linked list with a random pointer.
 * struct RandomListNode {
 *     int label;
 *     RandomListNode *next, *random;
 *     RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
 * };
 */
class Solution {
public:
    RandomListNode *copyRandomList(RandomListNode *head) {
        unordered_map<RandomListNode*, RandomListNode*> memo;
        auto dum = new RandomListNode(0);
        auto h = dum;
        while (head != nullptr) {
            if (memo.count(head) == 0) memo[head] = new RandomListNode(head->label);
            dum->next = memo[head];
            dum = dum->next;
            if (head->random != nullptr) {
                if (memo.count(head->random) == 0) memo[head->random] = new RandomListNode(head->random->label);
                dum->random = memo[head->random];
            }
            head = head->next;
        }
        return h->next;
    }
};

File: lc_java/d/Copy List with Random Pointer.java
/**
 * A linked list is given such that each node contains an additional random pointer which could point to any node in the list or null.
 * Return a deep copy of the list.
 */

/**
 * Definition for singly-linked list with a random pointer.
 * class RandomListNode {
 *     int label;
 *     RandomListNode next, random;
 *     RandomListNode(int x) { this.label = x; }
 * };
 */
public class Solution {
  public RandomListNode copyRandomList(RandomListNode head) {
    RandomListNode dum = new RandomListNode(0);
    RandomListNode pre = new RandomListNode(0);
    dum.next = head;
    Map<RandomListNode, RandomListNode> copied = new HashMap<>();
    while (dum.next != null) { // each iteraton clone next node: its label and its random
      dum = dum.next;
      RandomListNode n = copied.get(dum);
      if (n == null) {
        n = new RandomListNode(dum.label);
        copied.put(dum, n);
      }
      pre.next = n;
      pre = pre.next;
      if (dum.random != null) {
        RandomListNode r = copied.get(dum.random);
        if (r == null) {
          r = new RandomListNode(dum.random.label);
          copied.put(dum.random, r);
        }
        pre.random = r;
      }
    }
    return copied.get(head);
  }
}
